0001-short-term-release-process/0001-release-process-short-term.md

```diff
@@ -0,0 +1,325 @@
+
+In our setup, we would have two permanent branches, `main` and `stable`, where `main` is for
+day-to-day development and `stable` is essentially for tracking releases we consider deployable.
+Like Gitflow, we would also have some temporary branches; these will be discussed in more detail
+later.
+
+### Merging
+
+Historically, our team has preferred `git rebase` to `git merge`; however, Gitflow is better
+supported by using `git merge`. Rebasing is fine before submitting upstream, but Gitflow involves
+merging commits between different branches. Due to the fact that rebasing rewrites the commit
+history, merging between branches can lead to commits that have the same content but different
+hashes. This can make merging more confusing than it needs to be and also cause the commit history
+to be littered with duplicates. The primary option for completing a PR should be to merge it in
+rather than rebase.
+
+## Release Cycle Overview
+
+This section provides an overview of a release cycle. Certain aspects here merit more detail, but
+those will be provided in subsequent sections.
+
+The cycle has the following phases and steps:
+
+* Internal development and testing phase (two weeks):
+    - Begins immediately after production deploy of previous wave
+    - Deploy previous stable version to `STG-02` **
+    - Feature branches are worked on and merged back into `main`
+    - Developers can deploy their own isolated testnets if necessary
+    - We can do quick `alpha` releases if necessary
+* Release candidate (RC) phase (one week):
+    - Create a `release-YYYY.MM.rc.1` branch from `main`. No new features will be accepted on this
+      branch.
+    - Bump version numbers, with an `rc.1` suffix applied.
+    - Build RC and release to Github as a public `pre-release` but with no published crates.
+    - Deploy RC to `STG-01`.
+    - Production of changelog begins.
+    - Invite community members to test against this network.
+    - Shu's metrics solution enables comparisons to be made to the previous stable release in `STG-02`.
+    - Fixes can be applied to the release branch, resulting in an `rc.2`. This will be
+      released/deployed/tested. Repeat if necessary.
+* Release and deploy phase (one/two days):
+    - On the release branch, modify version numbers to remove the `rc` pre-release specifier.
+    - Changelog finalised.
+    - Merge the release branch into `stable`.
+    - Merge the release branch into `main`.
+    - When ready, perform a release and tag of `stable` by kicking off a workflow. *
+    - Delete the release branch.
+    - Deploy to `PROD-01`/`02` and drain *
+    - Announce to users
+
+We can accommodate hotfixes at any point during the cycle.
+
+#### Uncertainties
+
+* Not completely sure about the duration of each phase. Any opinions?
```

Comment:
I'd probably If we want to release to users for partial consumption (ie, say 5% of early adopters use the code), we'd probably want to go longer here to allow for this phase and subsequent assessment?

I'd say we want to be conservative here (and start so), with `alpha` or `beta` level code being released to the community before a final `rc`. (are all those required?)

Reply:
I think in a released network we would want to have monthly/bimonthly release candidate and longer testing period to ensure stabliilty, 

So perhaps we start faster (and no alphas) and expand into this flow over the coming months?

Reply:
Just to be clear, there is no `beta` in this proposal. The `alpha` releases may not happen, and if they do, they are intended for very quick experiments. Otherwise, it just goes to release candidate.

When you say "start faster" are you suggesting we should have a shorter development cycle than two weeks? Still one week for the RC testing phase?

Reply:
Ah, no i did not mean less than two weeks.

I mean: I think an ideal cadence in a live and functioning network will be many many weeks, maybe months. To be sure. (Bitcoin/eth is like one/6 months eg).

That is obviously quite long for us now. So perhaps we start with shorter cycles but increase them as we go?

(which would in turn offer time for alpha/beta releases with better burn in testing)

Reply:
Ah yeah, ok! I agree: I was initially thinking like 6 weeks. That then led me to narrowing the scope for this proposal to just cover the short term, which is for getting us to the initial release. Our process will probably be significantly different for post-release. I'm going to just go with what's stated in the proposal then, which is a three week cycle with two phases: two weeks for development, one week for RC testing, then deploy.